### PR TITLE
Read rulesets from AppDomain in addition to disk

### PR DESCRIPTION
**Pull request by miterosan**
This fixes the crash that happens in release config on android.

### CODE REVIEW

File: osu.Game/Rulesets/RulesetStore.cs [hunk 1]
```diff
@@ -22,9 +22,19 @@ static RulesetStore()
...
+            try
+            {
+                string[] files = Directory.GetFiles(Environment.CurrentDirectory, $"{ruleset_library_prefix}.*.dll");
+
+                foreach (string file in files.Where(f => !Path.GetFileName(f).Contains("Tests")))
+                    loadRulesetFromFile(file);
+            }
+            catch (Exception e)
+            {
+                Logger.Error(e, $"Could not load rulesets from directory {Environment.CurrentDirectory}");
```
Comment (peppy) [inline osu.Game/Rulesets/RulesetStore.cs]: isn't this going to trigger every startup on android?

File: osu.Game/Rulesets/RulesetStore.cs [hunk 2]
```diff
@@ -128,5 +151,20 @@ private static void loadRulesetFromFile(string file)
                 Logger.Error(e, $"Failed to load ruleset {filename}");
             }
         }
+
+        private static void addRuleset(Assembly assembly)
```
Comment (peppy) [inline osu.Game/Rulesets/RulesetStore.cs]: why is this function not being used in `loadRulesetFromFile`

**Comment by peppy** [comment]
Are you SURE this is the correct solution? Isn't the working directory of android still going to incorrectly be `/` ie. system root? Can you explain why this is the correct fix, and not instead potentially fixing the reason for this occurring in the first place?

image

This is currently done at the framework. It should guarantee the cwd is correct. If it isn't then we have larger issues, and should not be fixing it with local hacks.

**Comment by miterosan** [comment]
This is definitely required. There goes no way around this.

On android in Release configuration the assemblies are not being extracted to a directory.
Everything that the main Android project references will be loaded in the current appdomain.
There is no way to access the assemblies from the filesystem except if you write them from memory to the disk.

The working directory is set to where the application is started from. 

In debug configuration the shared mono runtime starts the application by running the assemblies that were copied into an hidden folder on the internal storage. The current directory "/" is in `/storage/sdcard0/android/data/ppy.osu.lazer/files/__override__/` because thats where the application is started from. The app has read and write rights there.

In Release configuration the apk is located in the ``/data/data/ppy.osu.lazer`` directory what equals to "/". The app does not have the right to enumerate the files here.
 
Setting the working directory to somewhere where the app does have read and write permissions does only fix the caught exception. It would not fix that the rulesets are not added to the rulesetStore.

**Comment by peppy** [comment]
Is there documentation on this (the chroot behaviour)? I tried searching but couldn't find much. Would be good to comment somewhere.

The larger issue is anything else expecting the working directory to be the app will also fail. If this is the case, we should check and remove all usages of it project-wide.

Also, we do not want to catch and log that exception if this is the case. It will be reported to us and the user every startup.

**Comment by miterosan** [comment]
There is not any documentation about this, except a few StackOverflow questions where they discuss this, sadly.

Currently the class RulesetStore is the only one accessing the current directory. (If you ignore osu.Framework.Platform.Architecture, which is windows only)

I think that logging it should not be a problem, as long ``Logger.Error`` is not used, right?

> Would be good to comment somewhere. 
Where would this information be appropriate to put?

**Comment by peppy** [comment]
I would write an inline comment in this code (in the `ctor`) mentioning the specific requirement of it existing for android.

**Review by peppy — APPROVED** [review]

**Pull request merged**